src/app/query-home/toolbar/query-page-header.tsx

```diff
@@ -0,0 +1,301 @@
+  const lakeId = useSelector(Current.getLakeId)
+  const searchBar = useSelector(SearchBar.getSearchBar)
+  const querySource = dispatch(getQuerySource(query?.id))
+  const [isEditing, setIsEditing] = useState(false)
+  const [isModified, setIsModified] = useState(false)
+
+  useEffect(() => setIsEditing(false), [query?.id])
+  useEffect(() => {
+    if (query?.value !== searchBar.current) setIsModified(true)
+    else setIsModified(false)
+  }, [query, searchBar])
+
+  const onSubmit = (newTitle) => {
+    setIsEditing(false)
+    if (newTitle !== "") {
+      query.name = newTitle
+      if (querySource === "draft") {
+        dispatch(Queries.addItem(query.serialize(), "root"))
+        dispatch(DraftQueries.remove({id: query.id}))
+        dispatch(tabHistory.replace(lakeQueryPath(query.id, lakeId)))
+      } else {
+        dispatch(updateQuery(query))
+      }
+    }
+  }
+
+  const renderQueryStatus = () => {
+    if (querySource === "draft" && !isEditing)
+      return (
+        <StyledAnchor isPrimary onClick={() => setIsEditing(true)}>
+          Save
+        </StyledAnchor>
+      )
+    let status = "Saved"
+    if (isModified) status = "Modified"
+    if (isEditing) status = "Renaming"
+    return (
+      <>
+        <span>{status}</span>
+        {!isEditing && <IconChevronDown />}
+      </>
+    )
+  }
+
+  const menu = usePopupMenu([
+    {
+      label: `Move to ${querySource === "local" ? "Remote" : "Local"}`,
+      click: () => {
+        if (querySource === "local") {
+          dispatch(setRemoteQueries([query.serialize()])).then(() => {
+            dispatch(refreshRemoteQueries())
+          })
+          dispatch(Queries.removeItems([query.id]))
+        } else {
+          dispatch(Queries.addItem(query.serialize(), "root"))
+          dispatch(setRemoteQueries([query.serialize()], true))
+        }
+      }
+    },
+    {
+      label: `Copy to ${querySource === "local" ? "Remote" : "Local"}`,
+      click: () => {
+        try {
+          const queryCopy = query.serialize()
+          queryCopy.id = nanoid()
+          if (querySource === "local") {
+            dispatch(setRemoteQueries([queryCopy]))
+          } else {
+            dispatch(Queries.addItem(queryCopy, "root"))
+          }
+          toast.success("Query Copied")
+        } catch (e) {
+          toast.error(`Copy Failed: ${e}`)
+        }
+      }
+    },
+    {
+      label: "Rename",
+      click: () => setIsEditing(true)
+    },
+    {
+      label: "Duplicate",
+      click: () => {
+        const dupeQ = query.serialize()
+        dupeQ.id = nanoid()
+        dupeQ.name += "-copy"
+        if (querySource === "local") {
+          dispatch(Queries.addItem(dupeQ, "root"))
+          dispatch(Tabs.new(lakeQueryPath(dupeQ.id, lakeId)))
+        }
+        if (querySource === "remote") {
+          dispatch(setRemoteQueries([dupeQ])).then(() => {
+            dispatch(Tabs.new(lakeQueryPath(dupeQ.id, lakeId)))
+          })
+        }
```

Comment:
It would be nice to have a single Query api for the CRUD operations that would find the source and do the right thing under the hood.

Reply:
Yeah, as this has grown I've had a similar thought

Reply:
Maybe this is belongs in the api.